File: src/BinaryTree/Exercise236.java
package BinaryTree;
public class Exercise236 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    class Solution {
        public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
            TreeNode left, right;

            if (root == null)
                return null;
            if (root == p || root == q)
                return root;

            left = lowestCommonAncestor(root.left, p, q);
            right = lowestCommonAncestor(root.right, p, q);

            if (left == null)
                return right;
            else if (right == null)
                return left;
            else
                return root;

        }
    }

    public static void main(String[] args) {
        Exercise236 exercise = new Exercise236();
        Solution solution = exercise.new Solution();
        TreeNode node6 = exercise.new TreeNode(7);
        TreeNode node5 = exercise.new TreeNode(4);
        TreeNode node4 = exercise.new TreeNode(2);
        TreeNode node3 = exercise.new TreeNode(6, null, node6);
        TreeNode node2 = exercise.new TreeNode(3, node4, node5);
        TreeNode node1 = exercise.new TreeNode(5, node2, node3);
        TreeNode test = solution.lowestCommonAncestor(node1, node2, node5);

    }
}

File: src/BinaryTree/Exercise236.cpp
/* Note:not solved by myself */

#include <iostream>
#include "BinaryTree.h"
using namespace std;

class Solution
{
public:
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q)
    {
        if (!root)
            return root;
        if (root == p || root == q)
            return root;
        TreeNode *left = lowestCommonAncestor(root->left, p, q);
        TreeNode *right = lowestCommonAncestor(root->right, p, q);
        if (left == nullptr && right == nullptr)
            return nullptr;
        else if (left == nullptr && right)
            return right;
        else if (left && right == nullptr)
            return left;
        else
            return root;
    }
};

int main()
{
    TreeNode a9(4);
    TreeNode a8(7);
    TreeNode a7(8);
    TreeNode a6(0);
    TreeNode a5(2, &a8, &a9);
    TreeNode a4(6, nullptr, nullptr);
    TreeNode a3(1, &a6, &a7);
    TreeNode a2(5, &a4, &a5);
    TreeNode a1(3, &a2, &a3);
    Solution sl;
    TreeNode *ans = sl.lowestCommonAncestor(&a1, &a2, &a3);
}